App multi-instance support

iamtrep opened the pull request:
Added section to allow renaming the app, to make management of multiple instances easier.  Different instances of the app can be necessary to log to different buckets, for example, or to turn off polling on some sensors.

Thread safety : changed @Field static variables to a Map type so that each app ID can have its own queue (and mutex)

Added parsing of "valve" event type to handleEvent() (works just like a switch)

iamtrep committed:
Update influxdb-logger.groovy

iamtrep committed:
Update influxdb-logger.groovy

dennypage commented:
Would you be so kind as to break this up into targeted changes to make review/merge easier please? Thanks

iamtrep commented:
I removed the offending commit.  Both commits left have the same purpose, i.e. allow multiple instances of the app to coexist.

dennypage commented:
FYI, here are the initial questions that come to mind:

* Should app name be optional? If so, why?
* Is the scoping of the variables in the definition section (loggerQueue and mutex) global to all apps? Sorry, I just don't know. I'm assuming it is, or the changes for these variables would not be needed. Have you tested to confirm?
* Why wouldn't the getMutex() and getLoggerQueue() functions be private?
* How are the global variables cleaned up?
* Can the mutex and queue variables be moved into state instead? This would obviate the need for global variables.
* What happens to an existing app if the new code is installed but app prefs are not re-saved? Is there a race condition? Is queued data lost?
* What happens if there are already multiple apps installed and the new code is installed without app prefs being re-saved?
* What happens if a second app with an app name is installed alongside an existing app that doesn't have a name?
* What happens if an app name is added to a previously running install? Is there a temporary race condition? Is queued data lost?
* Are there benefits to using a parent/child app structure instead?

iamtrep commented:
Great comments.  Thank you sir.  I'll see what I can address, especially the cleanup stuff, glad you pointed it out.

- The app name is actually just the app label.  Has no impact at all on functionality.  It's literally optional to change it but then the user will get a list of apps with the same name and that is harder to manage.  Probably useless to have the "(optional)" text in there since compulsory preferences are already marked by a *
- Scoping of a @Field static is apparently global from what I've read in the forums.  So the two Map objects are essentially  singletons (just like the current mutex and queue are right now).   I don't know if I can move them to state, I was having issues with dynamic typing, kinda hitting the limits of my groovy knowledge here.  But the situation is no worse than before, minus the cleanup point.
- Code update: there might be data loss on update if the current queue was not empty when the new code is installed.  I understand you are concerned with the upgrade path here.  I was obviously not, but you're totally right, I get it.  If you have any hints on how to handle this (e.g. leave old queue & mutex in there, empty queue prior to moving to new Map, etc) I'm all ears.
- Not sure I will go down the app parent/child route on this PR.  I'm reusing a pattern I've seen in a lot of popular Hubitat apps.  Nice idea though.

iamtrep committed:
improve per-instance thread safety objects & handling thereof

Using ConcurrentHashMap<> with appropriate types
Made getMutex() et al. private and moved to private helper function section.
Added a releaseLoggerQueue() private helper to write out any outstanding data before the thread-safety object of this app are released (called by uninstalled())

iamtrep committed:
removed (optional) from "Change Application Name"

iamtrep committed:
use appropriate labels for logger() log level

some calls to logger() with "Warning" as the log level (should be "warn")

iamtrep committed:
remove need for mutex

use atomic get-and-set methods from ConcurrentHashMap and ConcurrentLinkedQueue.

iamtrep commented:
So I made one last commit to remove the need for a mutex.  Simplifies the code a bit.  Doesn't require singleThreaded to be set to true.  Doesn't solve the new code load problem.

All things considered, if performance is not significantly worse, I think the singleThreaded version is simpler and better.

I propose to close this PR.

dennypage commented:
Okay. I've resubmitted the rebased state.loggerQueue changes in a new PR.
Thank you for all your work in bringing multi instance support to completion.
Greatly appreciated.
